cms/djangoapps/contentstore/helpers.py

```diff
@@ -345,6 +345,14 @@ def _import_xml_node_to_parent(
     else:
         for child_node in child_nodes:
             _import_xml_node_to_parent(child_node, new_xblock, store, user_id=user_id)
+
+    # ToDo: Check the better place to call it
+    # I tried to call it in the xml_block.py in the parse_xml() function,
+    # but the usage_key is not persited yet there
+    from cms.lib.xblock.tagging.tagged_block_mixin import TaggedBlockMixin
+    if isinstance(new_xblock, TaggedBlockMixin):
+        new_xblock.add_tags_from_xml()
+
```

Comment:
DragAndDrop fixed here: https://github.com/openedx/xblock-drag-and-drop-v2/pull/385

Reply:
@rpenido I don't know what was going on with my devstack yesterday, but I'm testing your branch again, and the issues I had yesterday aren't happening now. My apologies!  

> Is there some other XBlock that we should look for?

There's [a lot of XBlocks](https://openedx.atlassian.net/wiki/spaces/COMM/pages/43385346/XBlocks+Directory) that are variously maintained by Open edX or community members.. We can't possibly modify them all, and shouldn't have to either.. so I don't think we're headed down the correct path with https://github.com/openedx/xblock-drag-and-drop-v2/pull/385 and https://github.com/openedx/xblock-drag-and-drop-v2/pull/385.

I've raised https://github.com/open-craft/edx-platform/pull/640 to handle this issue better than we've been doing it. The issue is that TaggedBlockMixin is relying on the `xml_attributes` field from XmlMixin in order to do its thing, but not all XBlocks will be XmlMixins (since `XmlMixin` is not in [XBLOCK_MIXINS](https://github.com/openedx/edx-platform/blob/ea55eca47e81bd205b423373304c9437ef95915f/cms/envs/common.py#L973-L980)), and so we can't rely on this field being present.

Instead, I've added a `tags_v1` field to TaggedBlockMixin, and use that to store the serialized tag data. What do you think?

CC @yusuf-musleh @bradenmacdonald

Reply:
@pomegranited @rpenido 

Oh I see what's going on. So on `master` we have a [method](https://github.com/openedx/edx-platform/blob/master/cms/lib/xblock/tagging/tagged_block_mixin.py#L52-L57) in `TaggedBlockMixin` that would handle storing the tags in the xml for all  the "normal" blocks, since they all call the `add_xml_to_node` method, it overrides it and extends it. The HTML serialization was separate so we called the `add_tags_to_node` in there.

Seems like in this PR, we removed the `add_xml_to_node` method override from `TaggedBlockMixin`, hence we started seeing some inconsistent behavior, including blocks that do not have that method, because they would inherit it from `TaggedBlockMixin`. 

But I like @pomegranited's idea of having a new field that is separate from all that, and is more explicit on what it is and does, so it doesn't get accidentally removed cause some side effects.

Reply:
@pomegranited If the tags are stored in the field rather than the xml, how will it look in terms of importing/exporting? Will the tag data still be in the olx? The reason I ask is because I remember having a [discussion](https://tasks.opencraft.com/browse/FAL-3614?focusedCommentId=284882&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284882) with @bradenmacdonald about this, and in it we discussed how users may want to be able add/modify tag data manually in their course olx before importing for example.

I just wanted to check if this also covers the case for importing/exporting.

Reply:
I didn't test the course export/import case, but the tag data _should_ still be in the OLX same as it was before.
@rpenido  could you verify this?

Reply:
The OLX seems to be the same as before:
![image](https://github.com/openedx/edx-platform/assets/849463/4e82fe7a-f945-4a29-908e-ee037a3f4e11)

Thank you @pomegranited! You nailed it! I will create the PR reverting the changes in ORA, and I think we are good.

Reply:
ORA revert PR here: https://github.com/openedx/edx-ora2/pull/2184

Reply:
> Seems like in this PR, we removed the `add_xml_to_node` method override from `TaggedBlockMixin`

Why did we do that btw? Wasn't that working and (together with `XmlMixin`) it should cover almost all XBlocks? I don't think many XBlocks override the XML-related methods.

Reply:
I think the problem is that `XmlMixin` was not in [XBLOCK_MIXINS](https://github.com/openedx/edx-platform/blob/ea55eca47e81bd205b423373304c9437ef95915f/cms/envs/common.py#L973-L980), like @pomegranited mentioned. The way it is implemented now, it doesn't rely on the `XmlMixin` and doesn't need the changes in the specific XBlock package. Do you see a better way to do it?

Reply:
My concern with adding the tags as a field though, is now the tag data is going to be stored in modulestore/mongodb. So suddenly we have two sources of truth. And they could easily get out of sync. How do we know when exporting that it's using the latest tags (from the `content_tagging` API) and not the cached field data (from modulestore) from whenever the field was last updated?

I started thinking maybe we could use a `FieldOverrideProvider` or modify `CmsFieldData` to intercept this one field, and read its value from the tagging API rather than modulestore, but that would add a lot of overhead every time an XBlock is initialized or used in any way - but we only need this field during import/export.

So I'm leaning toward sticking with adding the tag field during `add_xml_to_node` and parsing it during `parse_xml`, without actually storing it as a field. I think that would work well for modulestore courses, since XML is only used during import/export anyways.

But now I realize that OLX/XML will also be used _all the time_ with Learning Core - not just for import/export but also for how the block data itself is "normally" stored. And I do feel that we probably don't want the tag data in the XML in that case (do we?) because again of the problem of two sources of truth and the need to keep it in sync. (Or do we want to allow people to change tags by editing the OLX and pressing save?)

Personally, I think the ideal would be to modify the platform so there is an extension point called for each block/node during import/export (only), to modify the OLX then.

@pomegranited @ormsbee @kdmccormick any thoughts on this? I'm feeling a bit uncertain here. It relates to a bunch of other stuff we've been discussing lately.

Reply:
@bradenmacdonald @rpenido 

> So I'm leaning toward sticking with adding the tag field during add_xml_to_node and parsing it during parse_xml, without actually storing it as a field.

Like this maybe? https://github.com/open-craft/edx-platform/pull/642/files#diff-9c6f63b94feb290ffbc18d01275713d04abc07e778fff0c42bc2a966a5c7dd6d

That PR also [adds a `studio_post_paste` handler](https://github.com/open-craft/edx-platform/pull/642/files#diff-4deb51bb5c3961a4a76ed1a264942c99df1cc63749fc6a614f8774ffe9266bcaR342) which works like the [`studio_post_duplicate` handler](https://github.com/search?q=repo%3Aopenedx%2Fedx-platform+studio_post_duplicate&type=code), so we don't have to check XBlocks to see if they're TaggedBlockMixins everywhere.

Reply:
@pomegranited Yes, that looks good to me :)

Reply:
Ongoing discussion on Slack too: https://openedx.slack.com/archives/C06JG6RFE4V/p1709667803595209

Reply:
@bradenmacdonald: I agree with your concerns about having multiple sources of truth. Given that this is one of those decisions that's likely to have long term repercussions, I want to make sure that we're all on the same page about why Learning Core is the way it is, and why that doesn't fit together well with the way that tagging is meant to work.

### Why store OLX?

XML is a little slow to parse and not particularly compact, so why should we store it instead of storing a more optimized format? Learning Core was designed in response to a few different problems we had with ModuleStore:

#### Exports should not need to rely on plugin code, e.g. an individual XBlock.

Content outlives software. When an XBlock is deprecated and removed, there is no way to know for certain what the serialization of that content data is supposed to look like. While most XBlocks just use the built-in defaults of having the fields expressed as attributes, there are XBlocks like ORA2 that give a much more complex and nested serialization.

This has caused issues in multiple cases where course export broke altogether for certain courses when these XBlocks were removed. The compromise behavior we have is that we now just skip trying to export block types we don't understand, but that's not a great solution, and we've seen multiple regressions in this behavior in the past.

#### Exports should be fast.

Invoking individual XBlocks at runtime to do serialization work can be slow and unpredictable. Invoking XBlock machinery also makes it memory intensive. Getting a simple dump of OLX from the database and writing it out to files will be much faster and require far less memory.

#### Exports should be robust.

When there is weird course content behavior on production, the first thing developers do is try to export a copy so that they can replicate the problem locally. Unfortunately, because the export process today requires executing a lot of XBlock plugin and runtime code, the error that's causing the erroneous behavior sometimes prevents the course from exporting altogether. At which point, trying to debug something is an infuriating exercise of trawling through the MongoDB shell on a read-replica of the prod server.

### How does Learning Core think about Truth?

As far as Learning Core is concerned right now, the true state of content are the immutable `ComponentVersions`, the `TextContents`/`FileContents` that they point to, and the `Draft` and `Published` information in the publishing app. So that's OLX and eventually other associated static assets.

Applications are encouraged to *extend* that representation with more optimized data models that are derivable from that raw OLX. So for instance, some XBlocks grab data from multiple "files", and it would be reasonable to make an `XBlockComponentVersion` model that was 1:1 with `ComponentVersion` and combined data from multiple `TextContents`. That might store the fields as JSON or something.

Another example I often use is that we might have an `ImageContent` that extends `FileContent` via a `OneToOneField` mapping, and gives metadata like width and height. We could also make these sorts of models for individual XBlocks, surfacing things like response types used, or video runtime.

These scenarios are like a progressive enhancement–they're not really adding new data to the system, but they're extracting the existing data in way that is far more useful. And we can get away with this safely because the raw data that they're enhancing stays immutable.

Whenever there's a new version, that OLX state gets written first, and everything is derived from it. We'll probably want to do that derivation synchronously, in the same transaction that we create the OLX. That way, we don't have to worry about things falling out of sync in case there's an error–the whole version creation gets rolled back.

### Why is Tagging different?

Everything I wrote above is about content itself. But tagging is living in a different lifecycle than content editing and publishing, which is where this all starts to fall apart.

As @bradenmacdonald  states in the Slack thread:

> tags live separately from the content publish history

Authors can add and remove a tag without ever publishing the component itself. Those tags immediately apply to every version of the component, whether it's been published or is still in draft form. The tags would not cause any change to the "last edited" or "last published" metadata associated with the component.

That's at the heart of Braden's point about this being a mess if it's stored in the OLX directly in Learning Core. Because if editing a tag means changing the OLX, then in Learning Core terms, that would require a new `ComponentVersion` pointing to OLX in a new `TextContent`. But tagging wants to change tagging data without ever touching those data models.

I would honestly much prefer that tagging did create new versions, so that tags are treated as peers of other attributes in the component OLX. It would be forcing tag annotations to live in the same lifecycle as content, which would simplify things like import/export. We'd have to have some code in edx-platform that was tagging aware enough to both save the appropriate attributes into the OLX and then write the matching data to the tagging API in the same transaction.

### So what now?

I don't have any great ideas. I would rather take the tags out of the attributes for individual component blocks, rather than to make it so that we have to dynamically filter them out and add them back in during export/import. Though I realize that introduces other problems around things like copy & paste.

I also worry about how tagging's decoupling from publishing means that the same library can give different results for queries based on tags, without any publishes/changes being recorded to anything inside of it. It's not enough to say, "did this collection change since the last time I checked?".

I'm also just apprehensive in ways that I can't properly articulate. One of the big headaches that we had around edx-val was trying to bridge import/export around things that were living very different lifecycles. I know that there are things that we can to do write it to a different store, but all of it just seems unnecessarily complex, and the end product is something that has more edge cases than if we just thought of tags as simple attributes with smarter post-processing.

I get that there are other types of tagging that we'll do eventually that don't live in the same lifecycle because they're made by people who don't have the same permissions–e.g. I make a personal tag for content I can't write to. But that kind of tagging wouldn't import and export with the content itself, as this does. The kind of tagging we're talking about now is editing done by the library author, that should respect import/export, and feels much cleaner to treat it more like the other attributes.

Reply:
@ormsbee Unfortunately we have already gone pretty far down this path of versioning tags separately and taking an "MVP" approach, which makes any major changes challenging. For example:

* Despite all the niceties of Learning Core's new API, we need to support tagging of (modulestore) course content today, and the idea of finding a way to make tag data fit into modulestore's versioning/publishing is highly unappealing.
* Taxonomies are themselves un-versioned. So even if a tag value is stored in the OLX of a particular ComponentVersion, if you then delete the taxonomy, you're left with a broken ComponentVersion that refers to a taxonomy tag that no longer exists. This isn't the end of the world, as the OLX likely includes either the taxonomy ID or name as well as the tag value, but it would be missing key data like the parent tags of that tag. If we want to bring tags into the publishing/versioning fold, I think we may end up having to do the same thing for taxonomies. (Which _could_ be nice long-term but introduces a lot of complexity at odds with our "MVP" approach, and creates challenges around import-export.)
* We've punted on the problem of global IDs for taxonomies with the concept of an "Export ID" - the "Export ID" of a taxnonomy is used only during import/export and it allows you to do cross-instance tag sharing in a way that's robust and easy to reason about. But to facilitate that, we allow changing the Export ID at any time, and we explicitly avoid using Export IDs for anything other than import/export. So the only stable identifier for a taxonomy suitable for use in OLX _within_ a given instance its its primary key (integer) which is not suitable for usage in _cross-instance_ OLX. Of course we could change the ID architecture if needed, but again - it's another layer of wrinkles.


I do have a couple ideas that may inspire further thoughts:
1. We could get rid of the idea of including tags in the OLX. Rather, when you export a course, we can export all the tags in a single `tag.json` file (or `.csv` or `.yaml` or `.xml` or whatever). (In fact, we already have a very similar [requirement](https://github.com/openedx/modular-learning/issues/172) so most of the work to support this is already done.) For copy-paste, it's easy to deal with tags separately from the OLX. However, I think this format is less friendly to power users who are editing the OLX. (That said, so is Studio's default export format in general)
2. We could stick with the current tagging API implementation but extend it so that ObjectTags (the link between tags and content) can optionally be applied to a range of ComponentVersion objects rather than just a Component. This still has some of the problems mentioned above with un-versioned taxonomies, but it keeps the tagging API generic, while still allowing tags to be partially versioned/published along with content in the way you're describing. Partially.

Reply:
I think that I'm casting my vote for (1) then: making it so that it's not embedded in the OLX. At the end of the day, the biggest thing that worries me is the mixing and magic of making something appear as part of the versioned content (OLX) when it's really metadata over the unversioned content, and the complexity that we'll face holding that facade together over time. Doing the import/export of that data separately may not be as pretty for power users, but I think it'll really help keep things simpler in the long term.

Reply:
One day, far, far, down the line, if we want to make a Taxonomy a PublishableEntity... 😉

Reply:
I'm leaning the same way now. Avoid the mismatch by not encoding tags in the OLX at all, and consider long-term options for leveraging PublishableEntity to make everything link together - but a probable prerequisite or corequisite for that is using Learning Core in courses.

Reply:
I'm curious how copy-and-paste is going to encode it though, if it's not in the OLX. Is there going to be some serialization of the tags that lives in staged content, kind of like how static assets are copied over?

Reply:
I was at first thinking of using an "on paste" event handler to just copy the tags over during paste. But I realize that doesn't support the "copy, delete, paste" workflow. So I think yeah, we'd add a new field to [`StagedContent`](https://github.com/openedx/edx-platform/blob/8a2c3ff14597e80be9440b8fe3d9cc8ca28cf451/openedx/core/djangoapps/content_staging/models.py#L29). It _could_ be a `tags` field but it's probably better to be a `extra_data` JSON field and then allow the `content_tagging` app to use that field to store/retrieve tags during copy/paste, so that the content_staging app doesn't need to be aware of tagging. Another option is to create a `StagedContentTags` model inside the `content_staging` app and use an "on copy" event handler to create the `StagedContentTags` model, and an "on paste" event handler to set the tags on paste.

Reply:
Actually, yet another idea requires no new models or DB fields at all. During a copy event, `ObjectTag` instances could be created that point to the `StagedContent` instance, rather than pointing to a `UsageKey`. I like this approach because the tags don't have to be serialized (they are represented in the DB the same way as other "normal" tags on content). The only downside is if the `StagedContent` contains child blocks, some arbitrary new ID format is required to indicate that tags apply to child blocks inside the StagedContent.

Reply:
@bradenmacdonald are you ok with the approach implemented here (for now)?

Reply:
Nudge @bradenmacdonald :)